code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
+            NVIC_SetPriority(iqrn, 9);
+            NVIC_EnableIRQ(iqrn);
 
             disable();
 
-            hardware_usart->US_BRGR = (5241600u / baudrate);
+            // write the baud rate in the BRGR register
+            detail::usart::port<Bus>->US_BRGR = (5241600u / baudrate);
 
-            hardware_usart->US_MR =
-                    UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
-            hardware_usart->US_IDR = 0xFFFFFFFF;
+            // set the usart mode
+            detail::usart::port<Bus>->US_MR =
+                UART_MR_PAR_NO | UART_MR_CHMODE_NORMAL | US_MR_CHRL_8_BIT;
 
             enable();
-        }
 
-        /// @brief char output operator
-        ///
-        /// Although calling send_byte should do the exact same thing.
-        /// In practice useing this fuction is more stable
-        /// Especially when the values are repeated.
-        hardware_usart_c &operator<<(uint8_t byte) {
-            send_byte(byte);
-            return *this;
+            // enable the interrupt for handling the receiving of data
+            detail::usart::port<Bus>->US_IER = US_IER_RXRDY;
         }
 
-        /// @brief string output operator
-        ///
-        /// used for char arrays to quickly send more than one byte
-        /// In practice useing this fuction is more stable than repeated use of
-        /// send_byte() Especially when the values are repeated.
-        hardware_usart_c &operator<<(const char *c) {
-            for (const char *p = c; *p != '\0'; p++) {
-                send_byte(*p);
+        /**
+         * @brief Write a uint8_t with the usart
+         *
+         * @param c data to send
+         * @return true if the data is sended correctly
+         * @return false if the data has not been sended correctly
+         */
+        bool send(const uint8_t c) override {
+            while (!transmit_ready()) {
+                // wait until transmit is ready
             }
-            return *this;
-        }
-
-        /// @brief enables the internal USART controller
-        void enable() override {
-            hardware_usart->US_CR = UART_CR_RXEN | UART_CR_TXEN;
-        }
 
-        /// @brief disables the internal USART controller
-        void disable() override {
-            hardware_usart->US_CR =
-                    UART_CR_RSTRX | UART_CR_RSTTX | UART_CR_RXDIS | UART_CR_TXDIS;
-        }
+            // write the byte in the Transmit holding register
+            detail::usart::port<Bus>->US_THR = c;
 
-        /// @brief send single byte
-        /// @param c: byte to send
-        /// @return true if byte send, false if not send
-        bool send(const uint8_t c) override {
-           send_byte(c);
-           return true;
+            return true;
         }
 
-        /// @brief recieve byte bia usart
-        /// @return received byte
+        /**
+         * @brief Get a data from the usart
+         *
+         * @return uint8_t
```

Comment:
Might be worth adding to the documentation (also doxygen since that's what the autocomplete is based on) that a `0` isn't necessarily the  data you might be expecting but rather the fact that there is no data at all.

Reply:
The early return should be removed as it should be undefined behavior when reading when no data is available

Reply:
If it's undefined behavior it'd be good to have it be explicity laid out in the readme/wiki

Reply:
I have added this as a `@warning` in the doxygen comments. Sorry for the user `@warning` I might have accidentally tagged you.